lib/ClangImporter/ClangImporter.cpp

```diff
@@ -1012,14 +1012,35 @@ ClangImporter::create(ASTContext &ctx,
 
   // Set up the file manager.
   {
-    if (!ctx.SearchPathOpts.VFSOverlayFiles.empty()) {
-      // If the clang instance has overlays it means the user has provided
-      // -ivfsoverlay options and swift -vfsoverlay options.  We're going to
-      // clobber their file system with our own, so warn about it.
-      if (!instance.getHeaderSearchOpts().VFSOverlayFiles.empty()) {
-        ctx.Diags.diagnose(SourceLoc(), diag::clang_vfs_overlay_is_ignored);
-      }
+    if (instance.getHeaderSearchOpts().VFSOverlayFiles.empty()) {
       instance.setVirtualFileSystem(ctx.SourceMgr.getFileSystem());
+    } else {
+      // Initialize the clang VFS from its compiler invocation.
+      instance.createFileManager();
+
+      // Create a new overlay file system for the clang importer with the clang
+      // VFS as its root.
+      auto ClangImporterFS =
+          llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>(
+              new llvm::vfs::OverlayFileSystem(
+                  instance.getVirtualFileSystemPtr()));
+
+      auto SwiftFS = ctx.SourceMgr.getFileSystem();
```

Comment:
Can we avoid changing the APIs from `llvm::vfs::FileSystem` by conditionally `dyn_cast`ing to `OverlayFileSystem` here?

Reply:
The VFS doesn't have the llvm RTTI thing, so not without adding that. Doing so would be relatively straightforward, but it goes against the sentiment of the VFS of staying agnostic of the implementation. I think such a change might be met with some resistance from the community.

Reply:
Ah, right, clients make their own FileSystems all the time. I suppose this is okay.